stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,316 @@
+        if (low == 0 || low > UINTPTR_MAX || address < (uintptr_t)low) {
+          continue;
+        }
+
+        unsigned long long high = strtoull(addrRange.second.str().c_str(),
+                                           nullptr, 16);
+        if (high == 0 || high > UINTPTR_MAX || address > (uintptr_t)high) {
+          continue;
+        }
+
+        auto fname = entry.split('/').second;
+        if (fname.empty() || fname.endswith(deleted)) {
+          continue;
+        }
+
+        fullPathName = "/" + fname.str();
+        break;
+      }
+      if (line) {
+        free(line);
+      }
+      fclose(fp);
+    }
+  }
+
+
+  // Parse the ELF binary using mmap for the section headers, symbol table and
+  // string table.
+  void mmapExecutable() {
+    struct stat buf;
+    int fd = open(fullPathName.c_str(), O_RDONLY);
+    if (fd < 0) {
+      return;
+    }
+
+    if (fstat(fd, &buf) != 0) {
+      close(fd);
+      return;
+    }
+
+    // Map in the section headers.
+    size_t sectionHeadersSize = (elfHeader->e_shentsize * elfHeader->e_shnum);
+    sectionHeaders = new Mapping(fd, buf.st_size, elfHeader->e_shoff,
+                                 sectionHeadersSize);
+    if (sectionHeaders->mapping) {
+      auto section = findSectionHeader(SHT_SYMTAB);
+      if (section) {
+        symbolTableData = new Mapping(fd, buf.st_size, section->sh_offset,
+                                      section->sh_size);
+        if (symbolTableData->mapping) {
+          symbolTable = section;
+        }
+      }
+      section = findSectionHeader(SHT_STRTAB);
+      if (section) {
+        stringTableData = new Mapping(fd, buf.st_size, section->sh_offset,
+                                      section->sh_size);
+        if (stringTableData->mapping) {
+          stringTable = section;
+        }
+      }
+    }
+    close(fd);
+    return;
+  }
+
+  // Find the section header of a specified type in the section headers table.
+  const Elf_Shdr *findSectionHeader(Elf_Word sectionType) {
+    if (sectionHeaders && sectionHeaders->mapping) {
+      auto headers = sectionHeaders->data<const Elf_Shdr *>();
+      for (size_t idx = 0; idx < elfHeader->e_shnum; idx++) {
+        if (idx == elfHeader->e_shstrndx) {
+          continue;
+        }
+        auto header = &headers[idx];
+        if (header->sh_type == sectionType) {
+          if (header->sh_entsize > 0 && header->sh_size % header->sh_entsize) {
+            fprintf(stderr,
+                    "section size is not a multiple of entrysize (%ld/%ld)\n",
+                    header->sh_size, header->sh_entsize);
+            return nullptr;
+          }
+          return header;
+        }
+      }
+    }
+    return nullptr;
+  }
+};
+
+
+int
+swift::lookupSymbol(const void *address, SymbolInfo *info) {
+  // The pointers returned point into the mmap()'d binary so keep the
+  // object once instantiated.
+  static auto binary = StaticBinaryELF();
```

Comment:
Instead of a static local, we prefer to use globals with the `Lazy` template, e.g.:

```c++
static Lazy<StaticBinaryElf> TheBinary;

int swift::lookupSymbol(const void *address, SymbolInfo *info) {
  auto &binary = TheBinary.get();
```

`Lazy` uses `dispatch_once`, which is a bit more efficient than `__cxa_guard`, and also ensures that a global destructor doesn't register. Global destructors run when the main thread exits, but `lookupSymbol` could be called from a thread that outlives the main thread.